### DISCUSSION TITLE
Setting an arbitrary property on API response

### DISCUSSION THREAD

**Discussion by vkuznecovas** [pocketbase/pocketbase#536] [Q&A]
Hello, 

This is an awesome project! I'm planning on using it as a framework for a project. I was wondering what would be the idiomatic way to enrich a record API response with an arbitrary property?

The first approach I tried involves using the `Expand`:
```go
	app.OnRecordViewRequest().Add(func(e *core.RecordViewEvent) error {
		if e.Record.TableName() != "my_collection" {
			return nil
		}

		r := e.Record.GetExpand()
		r["some_property"] = "some_value"
		e.Record.SetExpand(r)
		return nil
	})
```
However, this has the downside that the property is put under the `@expand` on the response. I'd probably have to also add a check to prevent overridint any references when queried with `expand` query param.

The second one involves setting the data:
```go
app.OnRecordViewRequest().Add(func(e *core.RecordViewEvent) error {
		if e.Record.TableName() != "my_collection" {
			return nil
		}

		// this requires `some_property` to be defined on the collection
		e.Record.SetDataValue("some_property", "some_value")
		return nil
	})
```
This carries the downside of having to define `some_property` as a field on the collection.

Which one of these would be the preferred approach? Perhaps this is something that could be added to the API to simplify this?

**Comment by ganigeorgiev** [comment]
The expand approach is ok, but to avoid collisions you have to prefix your property with `@` (field names aren't allowed to start with special characters).

Another approach, and probably the easiest, would be to just override the returned response in the event hook. For example (_I haven't tested it_):

```go
app.OnRecordViewRequest().Add(func(e *core.RecordViewEvent) error {
	if e.Record.TableName() != "my_collection" {
		return nil
	}

	data := e.Record.PublicExport()

	data["custom_prop"] = "test"

	e.HttpContext.JSON(http.StatusOK, data)

	return hooks.StopPropagation
})
```

Could you elaborate a little more on your use case and what type of metadata you want to append?

**Reply by vkuznecovas** [reply]
I use a `seed` property in the `collection` to generate a complex, pseudo-random object. 

Since the generation algorithm is stable I do not intend on storing the result itself in the `collection`. 

I want to avoid replication of the generation of the object on the client side, so I'd like to return the generated object as part of the API response.

**Reply by ganigeorgiev** [reply]
OK, I may consider adding support for serializing arbitary fields in the record json. I have some concerns that users may use it incorrectly and accidentaly leak some private only data, but I'll think more on this after the refactoring.

**Reply by vkuznecovas** [reply]
Thanks!

>  I have some concerns that users may use it incorrectly and accidentaly leak some private only data, but I'll think more on this after the refactoring.

I don't think this should be a particular concern as when using Pocketbase as a framework you already have the ability to do so by using the raw HTTP context, i.e. `e.HttpContext.String(200, "I just leaked a secret")`.
